File: algorithms/medium/design_front_middle_back_queue.ex
defmodule FrontMiddleBackQueue do
  defmodule DoubleLinkedList do
    defmodule Node, do: defstruct([:value, :prev, :next])
    defmodule LinkedList, do: defstruct([:size, :head, :tail])

    def new(), do: %ListedList{head: nil, tail: nil, size: 0}

    def is_empty?(list), do: list.size == 0

    def len(%LinkedList{head: _head, tail: _tail, size: size}), do: size

    def push_front(%LinkedList{head: nil, tail: nil, size: 0}, value) do
      new_node = %Node{value: value, prev: nil, next: nil}
      %LinkedList{head: new_node, tail: new_node, size: 1}
    end

    def push_front(%LinkedList{head: head, tail: tail, size: size}, value) do
      new_head = %Node{prev: nil, next: head}
      %LinkedList{head: new_head, tail: tail, size: size + 1}
    end

    def push_back(%LinkedList{head: nil, tail: nil, size: 0}, value) do
      new_node = %Node{value: value, prev: nil, next: nil}
      %LinkedList{head: new_node, tail: new_node, size: 1}
    end

    def push_back(%LinkedList{head: head, tail: tail, size: size}, value) do
      new_node = %Node{value: value, prev: tail, next: nil}
      %LinkedList{head: head, tail: new_node, size: size + 1}
    end

    def pop_front(%LinkedList{head: nil, tail: nil, size: 0} = list), do: {nil, list}

    def pop_front(%LinkedList{head: head, tail: nil, size: size}) do
      value = head.value

      if size == 1 do
        {value, %LinkedList{head: nil, tail: nil, size: 0}}
      else
        {value, %LinkedList{head: head.next, tail: tail, size: size - 1}}
      end
    end
  end

  defmodule Queue, do: defstruct([:left, :right])

  @spec init_() :: any
  def init_() do
  end

  @spec push_front(val :: integer) :: any
  def push_front(val) do
  end

  @spec push_middle(val :: integer) :: any
  def push_middle(val) do
  end

  @spec push_back(val :: integer) :: any
  def push_back(val) do
  end

  @spec pop_front() :: integer
  def pop_front() do
  end

  @spec pop_middle() :: integer
  def pop_middle() do
  end

  @spec pop_back() :: integer
  def pop_back() do
  end
end

# Your functions will be called as such:
# FrontMiddleBackQueue.init_()
# FrontMiddleBackQueue.push_front(val)
# FrontMiddleBackQueue.push_middle(val)
# FrontMiddleBackQueue.push_back(val)
# param_4 = FrontMiddleBackQueue.pop_front()
# param_5 = FrontMiddleBackQueue.pop_middle()
# param_6 = FrontMiddleBackQueue.pop_back()

# FrontMiddleBackQueue.init_ will be called before every test case, in which you can do some necessary initializations.

File: algorithms/medium/design_front_middle_back_queue.rs
use std::collections::VecDeque;

#[derive(Debug)]
struct FrontMiddleBackQueue {
    left: VecDeque<i32>,
    right: VecDeque<i32>,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl FrontMiddleBackQueue {

    fn new() -> Self {
        Self {
            left: VecDeque::new(),
            right: VecDeque::new(),
        }
    }
    
    fn push_front(&mut self, val: i32) {
        self.left.push_front(val);
        self.rebalance();
    }
    
    fn push_middle(&mut self, val: i32) {
        if self.left.len() == self.right.len() {
            self.left.push_back(val);
        } else if self.left.len() < self.right.len() {
            self.left.push_back(val);
        } else {
            let v = self.left.pop_back().unwrap();
            self.right.push_front(v);
            self.left.push_back(val);
        }

    }
    
    fn push_back(&mut self, val: i32) {
        self.right.push_back(val);
        self.rebalance();
    }
    
    fn pop_front(&mut self) -> i32 {
        if self.is_empty() {
            return -1;
        }
        
        if self.left.is_empty() {
            return self.right.pop_front().unwrap();
        }
        let v = self.left.pop_front().unwrap();
        self.rebalance();    
        v
    }
    
    fn pop_middle(&mut self) -> i32 {
        if self.is_empty() {
            return -1;
        }
        if self.left.len() >= self.right.len() {
            return self.left.pop_back().unwrap();
        }

        self.right.pop_front().unwrap()
    }
    
    fn pop_back(&mut self) -> i32 {
        if self.is_empty() {
            return -1;
        }
        if self.right.len() > 0 {
           let v = self.right.pop_back().unwrap();
            self.rebalance();
            return v; 
        }
        self.left.pop_back().unwrap()      
    }
    
    #[inline]
    fn is_empty(&self) -> bool {
        self.left.is_empty() && self.right.is_empty()
    }
    
    fn rebalance(&mut self) {
        if self.left.len() == self.right.len() {
            return;
        }
        if self.left.len() < self.right.len() {
            let v = self.right.pop_front().unwrap();
            self.left.push_back(v);
        } else if self.left.len() - 1 > self.right.len() {
            let v = self.left.pop_back().unwrap();
            self.right.push_front(v);
        }
    }
}

/**
 * Your FrontMiddleBackQueue object will be instantiated and called as such:
 * let obj = FrontMiddleBackQueue::new();
 * obj.push_front(val);
 * obj.push_middle(val);
 * obj.push_back(val);
 * let ret_4: i32 = obj.pop_front();
 * let ret_5: i32 = obj.pop_middle();
 * let ret_6: i32 = obj.pop_back();
 */
